refactor(devtools): save user settings

hawkgs opened the pull request:
**Work in progress**

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

- [x] Feature

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Settings are not saved after the tools are closed.

## What is the new behavior?

Introduces a new service for storing user settings that is used for saving the existing general settings.

All existing settings, apart from the theme, are now saved. The theme will be handled in a separate PR since the change is not trivial.

hawkgs commented on devtools/projects/shell-browser/src/app/chrome-application-operations.ts:
```diff
@@ -8,6 +8,8 @@
 
 /// <reference types="chrome"/>
 
+declare var browser: any;
```
JFYI: I tried installing `@types/firefox`. For some reason, `browser` doesn't exist.

JeanMeche replied:
Have you tried `@types/firefox-webext-browser` ?

hawkgs commented on devtools/src/local-storage.provider.ts:
```diff
@@ -0,0 +1,16 @@
+
+import {InjectionToken, Provider} from '@angular/core';
+
+export const LOCAL_STORAGE = new InjectionToken<typeof localStorage>('LOCAL_STORAGE');
+
+export const provideLocalStorage = (): Provider => ({
+  provide: LOCAL_STORAGE,
+  useValue: localStorage,
+});
```
I went with a simple injector rather than implementing a wrapper service since this will be purely used for the demo app; hence, dev purposes only.

JeanMeche replied:
we could provide a default factory to the token :
```suggestion
export const LOCAL_STORAGE = new InjectionToken<typeof localStorage>('LOCAL_STORAGE', { factory: () => localStorage });
```

This way the token behaves like a providedIn root service.

JeanMeche commented on devtools/projects/ng-devtools/src/lib/application-services/settings_store_service.ts:
```diff
@@ -0,0 +1,74 @@
+      }
+
+      runInInjectionContext(this.envInjector, () => {
+        effect(() => {
+          // Update the storage when the signal is updated.
+          this.data[key] = value();
+          this.appOperations.setStorageItems({[SETTINGS_STORE_KEY]: this.data});
+        });
+      });
```
This is the proper way to run an effect with an injector 
```suggestion
        effect(() => {
          // Update the storage when the signal is updated.
          this.data[key] = value();
          this.appOperations.setStorageItems({[SETTINGS_STORE_KEY]: this.data});
        });
      }, {injector: this.envInjector});
```

hawkgs replied:
`runInInjectionContext` has only this signature: 

```typescript
function runInInjectionContext<ReturnT>(injector: Injector, fn: () => ReturnT): ReturnT
```

... or I am missing something?

JeanMeche replied:
Sorry won't line, it's meant to be the option passed to the effect

hawkgs replied:
So, I guess we are good with the current approach (with the excl. of using the current injector instead of the env one)?

hawkgs replied:
Oh, okay I got you. I thought that `runInInjectionContext` is still required. We have to update our runtime error since it seems to be a bit outdated.

JeanMeche commented on devtools/projects/ng-devtools/src/lib/application-services/settings_store_service.ts:
```diff
@@ -0,0 +1,74 @@
+/** Provides an API for storing and preserving settings values. */
+export class SettingsStoreService {
+  private readonly appOperations = inject(ApplicationOperations);
+  private readonly envInjector = inject(EnvironmentInjector);
```
Why specifically an EnvironmentInjector here ?

JeanMeche commented on devtools/projects/ng-devtools/src/lib/application-services/settings_store_service.ts:
```diff
@@ -0,0 +1,74 @@
+   */
+  get<T>(key: string, initialValue: T): WritableSignal<T> {
+    // Create the signal with its initial value and return it synchronously.
+    const value = signal<T>(initialValue);
```
Shouldn't the signal be cached, so we don't create multiple signals for the same key ?

hawkgs replied:
Yeah, good point :+1:. I believe we can do it.

JeanMeche commented on devtools/projects/ng-devtools/src/lib/devtools.component.ts:
```diff
@@ -54,7 +55,7 @@ const LAST_SUPPORTED_VERSION = 9;
   ],
   imports: [DevToolsTabsComponent, MatTooltip, MatProgressSpinnerModule, MatTooltipModule],
-  providers: [WINDOW_PROVIDER, ThemeService],
+  providers: [WINDOW_PROVIDER, ThemeService, SettingsStoreService],
```
Providing at that level of the app is (almost) the same at providing in root. Shouldn't both `ThemeService` and `SettingsStoreService` be provided in root ?

JeanMeche commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/devtools-tabs.component.ts:
```diff
@@ -138,6 +145,13 @@ export class DevToolsTabsComponent {
       this.extensionVersion.set(chrome.runtime.getManifest().version);
     }
+
+    effect(() => {
+      const isDarkThemeEnabled = untracked(this.themeService.currentTheme) === 'dark-theme';
+      if (this.darkMode() !== isDarkThemeEnabled) {
+        this.themeService.toggleDarkMode(this.darkMode());
+      }
+    });
```
It feels like this `effect` is in the wrong location. Maybe should it be in the themeService itself ? 
We should keep a single source of truth for the theme, and right now with the storred setting, we're kinda diverging.

hawkgs replied:
I guess I can use the `SettingsStore` for the `currentTheme` signal instead. No need to move the effect. It will bind theme service to the settings service, but I assume it should be fine.

hawkgs commented:
I'll mark this as blocked until we get our local test environment in a working condition since introducing changes and, respectively, testing them at this moment is quite painful and time consuming.

hawkgs commented on devtools/projects/ng-devtools/src/lib/application-services/settings_store_service_spec.ts:
```diff
@@ -0,0 +1,149 @@
+  }));
+});
+
+class AppOperationsMock extends ApplicationOperations {
```
A leftover. It's already moved under `test-utils/settings_store_mock.ts`. I forgot to remove the copy :+1:

hawkgs commented on devtools/projects/ng-devtools/src/lib/application-services/settings_store_service.ts:
```diff
@@ -0,0 +1,74 @@
+    this.valuesLoaded.then(() => {
+      if (key in this.data) {
+        // When the data is loaded and if the key exists, update the signal value.
+        value.set(this.data[key] as T);
```
I reverted this from my initial iteration of the service since I think it's worth having the option to save objects and not limit the data only to primitive types. Also, there are some implications of constraining `T` and the TypeScript's typing system – type inference when using `get` is not working as desired. For example:

```typescript
get<T extends string | number | etc>(key: string, initialValue: T): WritableSignal<T> {}

const val = get('key@name', 'foo'); // => val: WritableSignal<'foo'>
```

So you would have to explicitly provide the template type like `get<string>('key@name', 'foo')`. Frankly, I haven't delved much into it, but these summarize the two reasons why `T` is not constrained.

dgp1130 replied:
> I reverted this from my initial iteration of the service since I think it's worth having the option to save objects and not limit the data only to primitive types.

IMHO, primitives would be fine here. I've gone down the road of serializing more complex objects and it just came out to a lot of effort for not much benefit. I'd suggest limiting this to primitives and then using more complex objects to wrap the setting in a `computed` or just wrap the whole signal as appropriate.

If you do think it's helpful to serialize complex objects, I'd suggest not implicitly `JSON.stringify`-ing them as this drops any methods in an unintuitive way. Rather, consider an explicit serialize / deserialize interface like [this](https://www.typescriptlang.org/play/?#code/PQKhAIGUFMCcEsCGAbeAvaBncjwBcBPAB2nABUA6cEYAKHgDs84AzRAY1Jjz0YHMYCFOjgAeMgD5wAb1rhwmOElQYAFAHsARgCsAXOQCU+zHgQM+AbjngAJliXC1iocmOn+R8lYC+tWqAgAGUYAa2xcdmRETGwAd3g8AAtwBOxnZRFYKhp6JlYOLmgefkEMtERNZGhxKVl5dMc4fW5ec1LG2BqfP2BgcABhRAYAcjwUgFsiKvHoJnAAAxaShxUKqvnbeFhodjxkAnBNHcQAV0UUvGG0vERedgAacF1VWkjo7DJE6BmZa3Z1BgmWAnXbqWCqIgIABut1I20QNgB+3AMOQJ2g+mGqD4iUu4AAPuBhjZELAQsMDDJfNYTLd4OxwPDEQxkQ0VE0oEVWgIVplxF8ZlIALy-eT1XlqJLfDHkAXQTxA-iisUq7Z4E6wBjgABSkAA8gA5CiK8zwFgEVTSKUzfTW6AUVHo7wGKwq7z3azyOxskSqNluMx8TyfaXKlXyNUarUMaCxWXS1SqXWGihEUmKP0rSnRHAMAgGCh2l2e8Du6zeKzWPhFABCfH66mQYNUCvc5jDCnieHYyVUSXgmAdKHRlLq4fk7GipCx8Bxl30kc1RNiiQS0GGrvH4En52JpPJC6KUaJlQ45M38l8l9o1N64AASlhG1DSFKFBK4LmbOAzqQEuAV1mRloARfhgFiBBuQoV4ohiTlinMTBIDwMFSDHeRqzwGpVAYRAbQUNs+EeH0MDWGUljaCUyJqTwW3AIUpDIDsVX+QExkwWd6IUWdcOQTMXBEMjjQ-LJvQ-VQmQovgcLw+UDGLEsVWgFgWB2PBE0pBjmK3CC1ykmSZmIj8hJIuBhIEpxZxbeSL3DZ1Ky3YD1SXDjLBLXxqTeOCG0mAFZjGdDwEheAYWYYCESRA4TFQrjGG0NTVCkpCUO2BSVTvfoqiGFI8lgNhOG3dRJjpSpSC7ZIvOwGYknUGxBxLJlIvwOUuP7Qdou2ChMNUUY5WGR4QxmYtVRA5lkU0etGzBLj-kmE5mBsDT6KkNrCzlFsutrSam3BGybz8Ow3m2cAWBOBhdngAEcvi3ZxHAaAAA9mAYOqZBjWIW30M6QgYdRYgYbwJFUN40DQfQyE8ABJNihk4MhiGqSQrEOqJjtO87eCu1zeOwxgEmUcHaM0xjkZ2VHSHRi6rtmoh5ugGxsPYTR9DorSIZZ4mvFoFHSQps6qa1ZTVN2YHmfAVmpChdR4BsTwpZl0mjr5jHLq1SSuX4AyZRNIMA34RXyZO-nMa1XTmH03D8J1x5HW1wi5elmwrCAA).

Like I said, it's a lot of effort to get right and probably isn't worth it unless and until we have a compelling use case which benefits from it.

> Also, there are some implications of constraining `T` and the TypeScript's typing system – type inference when using `get` is not working as desired.

Huh, I'm surprised you don't need `get('name', 'value' as const)` to get that effect. Checking with Gemini, I think you can either use function overloads or a type utility like:

```typescript
function get(name: string, initial: string): WritableSignal<string>;
function get(name: string, initial: number): WritableSignal<number>;
function get(name: string, initial: string | number): WritableSignal<string | number> {
  // ...
}

// OR

type Widen<T extends string | number> =
    T extends string
      ? string
      : T extends number
        ? number
        : string | number;
function get<T extends string | number>(name: string, initial: T): Widen<T>;
```

This is unfortunately more complicated than I'd really like it to be, but for a small enough set of primitives, maybe it's ok?

hawkgs replied:
Now that I started the refactoring, I realized that this approach is very restrictive and hard to extend. Let's take the themes, for example. We have a `Theme` string union type that we can't use and have to resort to plain string values that will need type assertions at some point in the code. 

The sacrificed flexibility outweighs the rare instances where a developer might make the mistake to save a non-serializable object in the store, in my opinion.

hawkgs replied:
Now that I started the refactoring, I've just realized that this approach is quite restrictive and hard to extend. Let's take the theme as an example. We have a `Theme` string union type that we can't really use or if we use via `Widen`, the return type will still be a `string`. This forces us to resort to type assertions in the code at some point.

The sacrificed flexibility outweighs the rare instances where a developer could make the mistake of storing a non-serializable data, in my opinion.

hawkgs commented on devtools/projects/ng-devtools/src/lib/application-services/settings_store_service.ts:
```diff
@@ -0,0 +1,74 @@
+
+@Injectable({providedIn: 'root'})
+/** Provides an API for storing and preserving settings values. */
+export class SettingsStoreService {
```
Sure. Though, I'll share my opinion on these:

1. Given this is a root service and technically a singleton, how could this be an issue? The service will live as long as the app lives. Or are you implying that since effects are dependent on change detection we don't have guarantee that the value will be updated? While I kind of get that point, I feel that in practice this is not a valid case. If I should pick an example (although it might not be the perfect one), it's like being concerned that you can't swim in an empty pool – yes, you won't be able to swim but you don't have the medium to begin with.
2. I didn't add a delete feature for now since there isn't a such use case and I couldn't see why we would need it in the foreseeable future. The service can be easily be extended when/if this is needed at all.
3. I think we can empirically agree that settings shouldn't be bothered with that much data. If large amounts of data need to be stored in the settings, then there is probably a design flaw somewhere else that needs to be addressed first.
4. Yes, I could put in place some sort of a buffering. Again, similarly to 3, if the settings data should be updated that frequently, you should probably go for a different storage solution.

dgp1130 replied:
For 1., I think the only practical way this happens is something calls `appRef.destroy()` (unlikely in this context) or the tab is closed. Basically if the user changes a setting, then kills the tab, it may not be persisted. Certainly not a significant issue.

I do agree these are all minor points and not something we need to spend too much time on, I more just wanted to state our assumptions out loud here.

hawkgs commented on devtools/projects/ng-devtools/src/lib/application-services/settings_store_service_spec.ts:
```diff
@@ -0,0 +1,149 @@
+    tick();
+
+    expect(foo()).toEqual('baz');
+    expect(bar()).toEqual('baz');
```
Yep :+1:

hawkgs commented on devtools/projects/ng-devtools/src/lib/application-services/theme_service.ts:
```diff
@@ -6,42 +6,45 @@
+  private readonly doc = inject(DOCUMENT);
+  private readonly settings = inject(SettingsStoreService);
+  private readonly renderer: Renderer2;
+  readonly currentTheme = this.settings.get<Theme>('theme@general', 'light-theme');
```
As for "theme in/at general settings". If we want to save the "Show full path" option in the "Router tree" tab, it would be named `show-full-path@router-tree`, for example. It's merely for being able to quickly identify the location of the settings (there is JSDoc comment/suggestion)>.

hawkgs replied:
> I don't personally see a ton of value in the category and would probably drop it entirely.

In addition of the extra context when inspecting the store directly, the `@location` postfix also prevents conflicts. I like the idea of separate context-based settings services but they leave the wrong impression that the keys are also context-specific. Let's say we want to save `sort_order` for feature `Foo` and feature `Bar` – having two separate services with `get('sort_order')` will lead to a bug, so you would still have to prefix or postfix the key somehow. Whether that's `sort_order_foo`, `sort_order_at_foo` `foo_sort_order` or `sort_order@foo` it doesn't really matter. The point is that `SettingsStore` is a store with unique keys after all and you are responsible for managing these keys, and dropping them, especially if we go with the services approach, is not a good idea, IMO.

Isn't adding a `category` unnecessarily complex?

hawkgs commented on devtools/projects/ng-devtools/src/lib/application-services/theme_service.ts:
```diff
@@ -6,42 +6,45 @@
 // Keep class names in sync with _theme.scss and _global.scss
-const DARK_THEME_CLASS = 'dark-theme';
-const LIGHT_THEME_CLASS = 'light-theme';
+export type Theme = 'dark-theme' | 'light-theme';
```
I introduced these class constants, so I am not really sure why I dropped them yesterday. I agree that classes shouldn't necessarily match the values. Should I just revert the change?

hawkgs commented on devtools/projects/ng-devtools/src/lib/application-services/settings_store_service_spec.ts:
```diff
@@ -0,0 +1,149 @@
+  let settingsStore: SettingsStoreService;
+  let getStoredSettings: () => {[key: string]: unknown};
+
+  beforeEach(fakeAsync(() => {
```
Thanks! On a different note, maybe this is indicative that `TestBed.tick` should be `TestBed.appTick`? I actually got confused that `tick` does what `TestBed.tick` does.

hawkgs commented on devtools/projects/ng-devtools/src/lib/application-services/settings_store_service.ts:
```typescript
// Describes the specific settings
interface DevtoolsSettings {
  'theme@general': string;
  'signal_graph@general': boolean;
}

// With or without a service, it seems that we'll need default/initial settings defined somewhere.
const DEFAULT_SETTINGS: DevtoolsSettings = {
  'theme@general': 'dark',
  'signal_graph@general': false,
};

// Provides an injector for `SettingsStore<DevtoolsSettings>`
export const SETTINGS = new InjectionToken<SettingsStore<DevtoolsSettings>>('SETTINGS');

export function provideSettings(): Provider[] {
  let savedSettings: Record<string, unknown> | undefined;

  return [
    provideAppInitializer(() => {
      const appOperations = inject(ApplicationOperations);
      savedSettings = await appOperations.getStorageItems(SETTINGS_STORE_KEY);
    }),
    {
      provide: SETTINGS,
      useFactory: () => {
        if (!savedSettings) throw new Error('App must initialize before loading settings.');
        return new SettingsStore<DevtoolsSettings>(savedSettings, DEFAULT_SETTINGS);
      },
    },
  ]
}

// SettingsStore types changes

class SettingsStore<T> {
  get<K extends keyof T>(key: K): T[K] {
    // ...
  }
}

// Usage

class Component {
  settings = inject(SETTINGS);

  // Where `get` accepts only keys of `DevtoolsSettings` and the return type
  // is automatically inferred from the interface.
  theme = this.settings.get('theme@general'); 
}
```
?? Or is it an overkill?

Ps. I guess this also highlights my preference for having a `get(key)` API 😄 (another perk is that the items are clearly distinguishable in the code from other services).

dgp1130 commented on devtools/projects/ng-devtools/src/lib/application-services/settings_store_service.ts:
Sure, I think having a version field and bumping that while running any necessary one-time migrations is potentially reasonable and roughly aligned with the way IndexedDB upgrading works.

Is there anything we should consider doing up front to prepare for that? In particular I'm thinking:
1. Write `version: 1` somewhere into the stored data. But we can maybe assume "If no version, default to 1", so I'm not sure how necessary this really is.
2. Some mechanism or convention to denote or protect data types are not safe to change without a migration. This would just be to protect ourselves from accidentally changing a value and breaking any existing stored values.

dgp1130 replied:
I'm happy to defer to you on what you think is useful to do at this stage, I mainly just wanted to make sure we were considering the use case and including any protections we think are useful up front.

dgp1130 commented on devtools/src/local-storage.provider.ts:
Style: In the interest of dropping suffixes, I'd probably call this just `local-storage.ts`.

dgp1130 commented on devtools/src/demo-application-operations.ts:
```diff
@@ -6,30 +6,87 @@
+
+  private setLsItems(items: object) {
+    try {
+      this.localStorage.setItem(STORAGE_KEY, JSON.stringify(items));
```
Note for privacy: While this is using Local Storage, it is limited to the demo application we're testing against. End user data is actually stored in `{browser,chrome}.storage.local`, not Local Storage.

dgp1130 reviewed (commented):
Launch tracking / privacy review: http://launch/4411656

hawkgs committed:
refactor(devtools): implement settings store

Introduces a set of services tasked with saving user settings.

hawkgs committed:
refactor(devtools): save user settings in the settings store

Save user settings in the `SettingsStore`.

NOTE: The theme is omitted since the change is not trivial and it will be handled in a separate PR.

hawkgs committed:
refactor(devtools): implement a migration mechanism for the settings store

Implement a simple mechanism that allows settings store data to be migrated, if needed.

hawkgs commented:
@dgp1130, all comments have been addressed. The change has been split into 3 commits:

1. Implementation of the store.
2. Integration with the front-end app (i.e. existing settings are saved). Note that the theme is not saved since the change isn't trivial and will add too much additional noice. I'll create a follow-up PR after this one is merged.
3. Implementation of a simple migration mechanism **(NEW)**.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/application-providers/settings_provider.ts:
```diff
@@ -11,14 +11,20 @@ import {SETTINGS_STORE_KEY, SettingsStore} from '../application-services/setting
       const keyedItem = await appOperations.getStorageItems([SETTINGS_STORE_KEY]);
-      savedSettings = (keyedItem[SETTINGS_STORE_KEY] ?? {}) as {[key: string]: unknown};
+      const data = (keyedItem[SETTINGS_STORE_KEY] ?? {}) as {[key: string]: unknown};
+      applyMigrations(data, appOperations);
```
Consider: Have this return the migrated data rather than mutate the `data` parameter to be more intuitive and reduce the possibility of unexpected mutations.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/application-providers/settings_provider.ts:
```diff
@@ -27,3 +33,15 @@ export function provideSettings(): (Provider | EnvironmentProviders)[] {
   ];
 }
+
+function applyMigrations(data: {[key: string]: unknown}, appOperations: ApplicationOperations) {
```
Consider: It would be helpful to include a detailed doc comment explaining what we mean by "migration" in this context and how it should be used.

As it is right now, this is basically just dead code and an isolated observer would probably struggle to understand it. So some documentation or an example of a migration would be helpful for when we need to actually use it.

dgp1130 reviewed (commented):
LGTM on the implementation front with only some minor (non-blocking) suggestions.

Just waiting on the launch approval, hoping to get that resolved by tomorrow.

hawkgs commented on devtools/projects/ng-devtools/src/lib/application-services/test-utils/settings_mock.ts:
```diff
@@ -0,0 +1,22 @@
+import {Provider, signal, WritableSignal} from '@angular/core';
+import {Settings} from '../settings';
+
+export class SettingsMock {
```
We can't `implement` it due to the private members. I had initially opted out of `extend` since it introduces `SettingsStore` as a dependency, but now I've reverted that and created a `SettingsStore` mock as well. The mock doesn't extend the `SettingsStore` per se, but at least `SettingsMock`, which has the potential to change, is now type safe.

hawkgs committed:
fixup! refactor(devtools): save user settings in the settings store

hawkgs committed:
fixup! refactor(devtools): implement a migration mechanism for the settings store

hawkgs committed:
refactor(devtools): save transfer state tab state in the settings store

Save the Transfer State tab state in the Settings store.

dgp1130 reviewed (approved):
Got the relevant approvals, so we're good to rebase, merge, and release this whenever we're ready.

devversion reviewed (approved):
LGTM for dev-infra
